### DISCUSSION TITLE
XrmRealContext

### DISCUSSION THREAD

**Discussion by victorsolaya** [DynamicsValue/fake-xrm-easy#34] [Q&A]
Hi, I have been playing a little bit with the new 2.1.0 release version, however I have some doubts about the XrmRealContext. 
Before in the v1 you could do a xrmRealContext.getDefaultPluginContext(), however that option in the v2 doesnt appear.
How is supposed to work this feature in order to perform IntegrationTesting?

Many thanks for the help!

**Comment by jordimontana82** [comment]
Hi @victorsolaya , interesting question.

Why do you need that method in the XrmRealContext class? It's indeed there, mostly because of being inherited from a common interface but I never used it nor it's documented in that way.

What I did find useful sometimes, for troubleshooting, was the [GetPluginContextFromSerialisedProfile](https://github.com/DynamicsValue/fake-xrm-easy-core/blob/f343198c49c5eb68b1940173c7668b2505cf5b75/src/FakeXrmEasy.Core/XrmRealContext.cs#L170), which allows you to replay a plugin execution after a serialised, compressed plugin execution context you can get from the plugin profiler. But this wasn't intented for integration testing, mostly to debug a plugin against a target environment to troubleshoot something.

As per integration testing, you can always call methods on the IOrganizationService that will make any plugins trigger (i.e. service.Create() will trigger any plugins registered against Create message), that's what I would do for integration testing if needed.

Although I mostly heavily rely on unit testing alone along with  [the new pipeline simulation capability to execute chains of plugin executions In-Memory alone](https://dynamicsvalue.github.io/fake-xrm-easy-docs/releases/2x/2.1.0/) sometimes.

**Reply by victorsolaya** [reply]
Hi Jordi,

The idea behind the integration testing was found based on this post https://dreamingincrm.com/2016/11/21/plugin-integration-tests-using-fakexrmeasy/ 
I really liked the idea of having everything in testing as you can perform (or I believe is much better) to have everything in Asserts to check results with Custom APIs or even with Plugins checking against the Service.

My question was based on the fact I couldn't find that method that was on v1, I didn't know if you had any plans to implement it at some point, I believe they could be really helpful at some point, or that's what I think but your point is quite good as well.

Thank you!

**Reply by jordimontana82** [reply]
You can still have Asserts using a real service without that method, with even less boilerplate (just a single line of code).

What's the use case? I mean, my point is, by the time you are done with unit testing the plugin there is not much left for integration testing in this context I mean....

Do you want just to debug the plugin code locally but connected to a real environment, maybe?

**Reply by victorsolaya** [reply]
Yes, exactly, because sometimes installing the profiler is not the best option as it sometimes fails 😞

**Reply by jordimontana82** [reply]
Ok, added it to the next release: #35

**Reply by victorsolaya** [reply]
That is awesome. Many thanks for the help Jordi!

**Reply by jordimontana82** [reply]
No worries
